commons-test/src/main/java/org/infinispan/commons/test/RunningTestsRegistry.java

```diff
@@ -0,0 +1,96 @@
+   private static final Map<Thread, ScheduledFuture<?>> scheduledTasks = new ConcurrentHashMap<>();
+
+   static void unregisterThreadWithTest() {
+      ScheduledFuture<?> killTask = scheduledTasks.remove(Thread.currentThread());
+      killTask.cancel(true);
+   }
+
+   static void registerThreadWithTest(String testName, String simpleName) {
+      Thread testThread = Thread.currentThread();
+      ScheduledFuture<?> future = executor.schedule(
+         () -> killLongTest(testThread, testName, simpleName), MAX_TEST_SECONDS, SECONDS);
+      scheduledTasks.put(testThread, future);
+   }
+
+   @SuppressWarnings("deprecation")
+   private static void killLongTest(Thread testThread, String testName, String simpleName) {
+      try {
+         String safeTestName = testName.replaceAll("[^a-zA-Z0-9=]", "_");
+         System.err.printf(
+            "Test %s has been running for more than %d seconds. Interrupting the test thread and dumping " +
+               "thread stacks of the test suite process and its children.\n",
+            testName, MAX_TEST_SECONDS);
+
+         String jvmName = ManagementFactory.getRuntimeMXBean().getName();
+         String ppid = jvmName.split("@")[0];
+         ArrayList<String> pids = new ArrayList<>(singleton(ppid));
+         int index = 0;
+         while (index < pids.size()) {
+            String pid = pids.get(index);
+            Process ps = new ProcessBuilder()
+               .command("ps", "-o", "ppid=", pid)
+               .start();
+            BufferedReader psOutput = new BufferedReader(new InputStreamReader(ps.getInputStream()));
+            psOutput.lines().forEach(line -> pids.add(line.trim()));
+            ps.waitFor(10, SECONDS);
+
+            File dumpFile =
+               new File(String.format("threaddump-%1$s-%2$tY-%2$tm-%2$td-%3$s.txt", safeTestName, new Date(), pid));
+            System.out.printf("Dumping thread stacks of process %s to %s\n", pid, dumpFile.getAbsolutePath());
+            Process jstack = new ProcessBuilder()
+               .command(System.getProperty("java.home") + "/../bin/jstack", pid)
+               .redirectOutput(dumpFile)
+               .start();
+            jstack.waitFor(10, SECONDS);
+         }
+
+         // Interrupt the test thread
+         testThread.interrupt();
+         System.out.printf("Interrupted thread %s (%d).", testThread.getName(), testThread.getId());
+
+         testThread.join(SECONDS.toMillis(MAX_TEST_SECONDS) / 10);
+         if (testThread.isAlive()) {
+            // Thread.interrupt() doesn't work if the thread is waiting to enter a synchronized block or in lock()
+            // Thread.stop() works for lock(), but not if the thread is waiting to enter a synchronized block
+            // So we just kill the fork and its children instead
+            // Ignore the error on Windows for now
+            Process kill = new ProcessBuilder()
+               .command("kill " + String.join(" ", pids))
```

Comment:
This obviously won't work for Windows. Shouldn't we detect the OS and just skip this whole mechanism in that case?

Reply:
Yeah, I'll do that.